src/main/flight/gps_rescue.c

```diff
@@ -606,11 +607,11 @@ void updateGPSRescueState(void)
                 newAltitude = gpsRescueConfig()->initialAltitudeM * 100;
                 break;
             case CURRENT_ALT:
-                newAltitude = rescueState.sensor.currentAltitudeCm;
+                newAltitude = rescueState.sensor.currentAltitudeCm + gpsRescueConfig()->rescueAltitudeBufferM * 100;
```

Comment:
Not sure this is the right thing to do - it changes the behaviour of the `CURRENT_ALT` setting as well as the default behaviour, which will be unexpected for users.
More generally, if a buffer is wanted in this case, should it be applied to all cases?

Reply:
>it changes the behaviour of the CURRENT_ALT setting as well as the default behaviour
I wasn't sure on this either and shared a similar concern; I was mostly trying not to just come up with more options or complicate things too much with an extra method, but I definitely see the argument for just making a new method and changing nothing about existing methods.

- Would you advise that I just create a new method? "CURR_ALT_W_BUFFER"?

> More generally, if a buffer is wanted in this case, should it be applied to all cases?
I just wasn't sure about the development flow you had in mind. I felt the changes were minor (the default setup won't be changed if you leave the buffer untouched) and would prevent further development of more options like "CURR_ALT_W_BUFFER", "MAX_ALT_W_BUFFER" as time goes on (since MAX_ALT already has one hard coded anyways).

Happy to change as advised. Just let me know how we can make the decision.

Reply:
> Not sure this is the right thing to do - it changes the behaviour of the `CURRENT_ALT` setting as well as the default behaviour, which will be unexpected for users.

@mikeller 
would it then be appropriate to simply do nothing about "MAX_ALT" and introduce the CLI variable "gps_rescue_alt_buffer" only for the "CURRENT_ALT" mode and set it to "0" as a default value ?
then nothing changes for the existing users even if they are unaware this improvement was implemented.
I suspect the users who become aware if this change may never use MAX_ALT again anyway. the only reason people are still using "MAX_ALT" today is because the alternative "CURRENT_ALT" has no safety buffer/margin.

Reply:
oh, actually @peteoz, that's a great idea. perhaps that's the best implementation.

Reply:
@peteoz: So your suggestion is to only modify `CURRENT_ALT`, and leave `MAX_ALT` with the hardcoded buffer? That's one way to do it. But if it is the case that many users want to use `CURRENT_ALT` but with a buffer, then making the buffer configurable in `gps_rescue_alt_buffer`, and adding it to both will lead to the most desirable outcome - and it will require a configuration change from these users anyway.

Reply:
@mikeller So it is sounding more like you'd want a new method then?

Would acceptable changes include:

1. `gps_rescue_alt_buffer` = default at 15m
2. `CURRENT_ALT_BUFFER` = Current altitude + buffer in all cases. **new method**
3. `MAX_ALT` = existing behavior, but after initial altitude is hit, `gps_rescue_alt_buffer` kicks in instead of 15m fixed buffer.
4. `CURRENT_ALT` = leave unchanged? with `gps_rescue_alt_buffer` = 0 it would be the same as CURRENT_ALT_BUFFER.

Reply:
Na, I'd roll 2. and 4. into `CURRENT_ALT` = Current altitude + buffer in all cases. I think that the point that @peteoz makes that currently `CURRENT_ALT` is problematic for users is valid - returning without ascending first will even crash into bystanders on the way (worst case). We'll just need to make sure to mention the change in behaviour for current users of `CURRENT_ALT` in the release notes.

Reply:
@mikeller that suggestion was only in order to preserve current behaviour for all existing users because there is no other way around it. under normal circumstances (eg. if these features were implemented for the first time) my suggestion would be to have configurable variable "gps_rescue_alt_buffer" (default at 15m) applicable to both MAX_ALT and CURRENT_ALT and then just forget about "CURRENT_ALT_BUFFER". but then of course once implemented the CURRENT_ALT will behave differently in that there'll be the initial ascend of 15m that existing users don't expect. in my opinion this "safety" ascend is not detrimental and still it gets mentioned in the released notes. so to summarise - my preferred course of action would be simply to introduce "gps_rescue_alt_buffer" (default at 15m) and make it applicable to both MAX_ALT and CURRENT_ALT, then mention in the release notes that for mode CURRENT_ALT there'll be initial ascend before returning.

Reply:
@peteoz: I think we mean the same. :-)

Reply:
Seemingly in consensus @mikeller & @peteoz - updated the code to hopefully reflect the requested changes. Built and flashed a target - I'll try to test today. If things are successful, I'll post some builds here.

Reply:
@peteoz, FYI - ran a test yesterday that worked out well with these changes. Even moderate testing under these circumstances made it clear that 15m is, as you said, likely the bare minimum you'd want. Added some test firmware above with mikeller's installation guide above if you wanted to try it out. thanks for helping out with this!

Reply:
@EggsBenedict great, I installed test FW on 2 of my quads, going to test this week if the weather is kind, thanks for all the coding. not sure if this is the right place to ask/discuss but I was wondering if you happen to know what is the default behaviour for "CURRENT_ALT" mode when the quad is flown deep below the launch site (eg. from a ridge down into a valley -200m below) and then failsafe, for example - would "CURRENT_ALT" reset itself to equivalent of launch site altitude ? I guess what I'm asking - whether you know how was the original version coded for this eventuality.

Reply:
@peteoz, perhaps this keys into what @mikeller was getting at to certain degree. My understanding is that - no, it would stay low.  The way that `MAX_ALT` is setup you would be insulated from this and in such a way that be applied to `CURRENT_ALT` pretty easily if you think this is a case that should be caught. The logic is getting slightly complicated though from a user standpoint. I sort of wonder if a separate PR should be setup for a new method like `POSITIVE_CURRENT_ALT`? But, also, I mean to me it makes a lot of sense to include it in this method. Perhaps @mikeller would have guidance on this.

`initialAltitudeM` = user configurable min safe altitude (default = 50m)

`MAX_ALT` takes the higher of the following:
- `initialAltitudeM`
- `maxAltitudeCm` + `rescueAltitudeBufferM`

You could update `CURRENT_ALT` to take the higher of the following:
- `initialAltitudeM`
- `currentAltitudeCm` + `rescueAltitudeBufferM`

And that would catch this use case of diving into a valley below the take off altitude.

Reply:
@EggsBenedict , I was only asking so that I know what to expect, but I think current implementation is acceptable. otherwise we would be getting back to the scenario of "killing the battery on big ascend". under normal circumstances I would still expect to regain the signal as soon as the quad ascends from behind the obstacle. I suppose if there is someone who always or often flies below the launch site that such user would actually keep MAX_ALT as the rescue mode. ideally at some point the BF LUA script gets updated so that we can switch between different rescue modes directly on the radio before the flight, and in fact I would also like the new variable "gps_rescue_alt_buffer" to become available in the BF LUA script.

Reply:
sounds good! i don't mess with LUA too much - but the rescue modes are in the OSD menus i think so you can change them on the fly easily. as per mikeller's request, changing the buffer value won't be incorporated into the OSD immediately, but eventually if the change sticks:
> it should not be added until the new parameter has been included in at least one release, to avoid churn in MSP due to changing parameters.
which makes sense.

Reply:
> @peteoz, FYI - ran a test yesterday that worked out well with these changes. Even moderate testing under these circumstances made it clear that 15m is, as you said, likely the bare minimum you'd want. Added some test firmware above with mikeller's installation guide above if you wanted to try it out. thanks for helping out with this!

@EggsBenedict I ran my tests with the test firmware yesterday, using CURRENT_ALT with gps_rescue_alt_buffer = 50, everything worked as expected.

Reply:
Awesome! Thanks for testing @peteoz! Did a few more tests on the F7X2 target on a 7" - also worked as expected.